oneandone/src/main/java/org/apache/jclouds/oneandone/rest/compute/strategy/AssignDataCenterToTemplate.java

```diff
@@ -0,0 +1,89 @@
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Resource;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import org.apache.jclouds.oneandone.rest.OneAndOneApi;
+import org.apache.jclouds.oneandone.rest.domain.DataCenter;
+import static org.jclouds.Constants.PROPERTY_USER_THREADS;
+import org.jclouds.compute.config.CustomizationResponse;
+import org.jclouds.compute.domain.NodeMetadata;
+import org.jclouds.compute.domain.Template;
+import org.jclouds.compute.functions.GroupNamingConvention;
+import org.jclouds.compute.reference.ComputeServiceConstants;
+import org.jclouds.compute.strategy.CreateNodeWithGroupEncodedIntoName;
+import org.jclouds.compute.strategy.CustomizeNodeAndAddToGoodMapOrPutExceptionIntoBadMap;
+import org.jclouds.compute.strategy.ListNodesStrategy;
+import org.jclouds.compute.strategy.impl.CreateNodesWithGroupEncodedIntoNameThenAddToSet;
+import org.jclouds.logging.Logger;
+
+@Beta
+@Singleton
+public class AssignDataCenterToTemplate extends CreateNodesWithGroupEncodedIntoNameThenAddToSet {
+
+   @Resource
+   @Named(ComputeServiceConstants.COMPUTE_LOGGER)
+   protected Logger logger = Logger.NULL;
+
+   private final OneAndOneApi api;
+
+   @Inject
+   protected AssignDataCenterToTemplate(
+           CreateNodeWithGroupEncodedIntoName addNodeWithGroupStrategy,
+           ListNodesStrategy listNodesStrategy,
+           GroupNamingConvention.Factory namingConvention,
+           @Named(PROPERTY_USER_THREADS) ListeningExecutorService userExecutor,
+           CustomizeNodeAndAddToGoodMapOrPutExceptionIntoBadMap.Factory customizeNodeAndAddToGoodMapOrPutExceptionIntoBadMapFactory,
+           OneAndOneApi api) {
+      super(addNodeWithGroupStrategy, listNodesStrategy, namingConvention, userExecutor,
+              customizeNodeAndAddToGoodMapOrPutExceptionIntoBadMapFactory);
+      this.api = api;
+   }
+
+   @Override
+   public Map<?, ListenableFuture<Void>> execute(String group, int count, final Template template,
+           Set<NodeMetadata> goodNodes, Map<NodeMetadata, Exception> badNodes,
+           Multimap<NodeMetadata, CustomizationResponse> customizationResponses) {
+
+      logger.info(">> looking for a datacenter in %s", template.getLocation().getId());
+
+      // Try to find an existing datacenter in the selected location
+      DataCenter dataCenter = find(api.dataCenterApi().list(), new Predicate<DataCenter>() {
+         @Override
+         public boolean apply(DataCenter input) {
+            // The location field is not populated when getting the list of datacenters
+            DataCenter details = api.dataCenterApi().get(input.id());
+            return details != null && template.getLocation().getId().equals(details.countryCode().toLowerCase());
+         }
+      });
+
+      return super.execute(group, count, new TemplateWithDataCenter(template, dataCenter), goodNodes, badNodes,
+              customizationResponses);
+   }
```

Comment:
If we already use the DataCenter object as the location object in the adapter, we shouldn't need this class, as this list would be performed in the list locations method.